File: src/DO/Graphics/DerivedQObjects/PaintingWindow.cpp
// ========================================================================== //
// This file is part of DO++, a basic set of libraries in C++ for computer 
// vision.
//
// Copyright (C) 2013 David Ok <david.ok8@gmail.com>
//
// This Source Code Form is subject to the terms of the Mozilla Public 
// License v. 2.0. If a copy of the MPL was not distributed with this file, 
// you can obtain one at http://mozilla.org/MPL/2.0/.
// ========================================================================== //

#include "PaintingWindow.hpp"
#include <QtWidgets>

namespace DO {

	// ====================================================================== //
	// ScrollArea
	ScrollArea::ScrollArea(QWidget* parent) : QScrollArea(parent)
	{
    setAlignment(Qt::AlignCenter);
	}

	void ScrollArea::closeEvent(QCloseEvent *event)
	{
    if(event->spontaneous())
    {
      qWarning() << "\n\nWarning: you closed a window unexpectedly!\n\n";
      qWarning() << "Graphical application is terminating...";
      qApp->exit(0);
    }
	}

	// ====================================================================== //
	// PaintingWindow
  PaintingWindow::PaintingWindow(int width, int height,
                                 const QString& windowTitle, int x, int y,
                                 QWidget* parent)
    : QWidget(parent)
    , scroll_area_(new ScrollArea(parent))
    , pixmap_(width, height)
    , painter_(&pixmap_)
  {
    resize(width, height);
    //setMouseTracking(true);
    setFocusPolicy(Qt::WheelFocus);

    event_listening_timer_.setSingleShot(true);
    connect(&event_listening_timer_, SIGNAL(timeout()),
      this, SLOT(eventListeningTimerStopped()));

    if(x != -1 && y != -1)
      scroll_area_->move(x,y);
    scroll_area_->setWindowTitle(windowTitle);
    scroll_area_->setWidget(this);
    scroll_area_->setFocusProxy(this);

    if (width > qApp->desktop()->width() || height > qApp->desktop()->height())
    {
      width = 800;
      height = 600;
    }

    scroll_area_->resize(width+2, height+2);

    pixmap_.fill();
    update();

    scroll_area_->show();
  }

	void PaintingWindow::drawPoint(int x, int y, const QColor& c)
	{
		painter_.setPen(c);
		painter_.drawPoint(x, y);
		update();
	}

  void PaintingWindow::drawPoint(const QPointF& p, const QColor& c)
  {
    painter_.setPen(c);
    painter_.drawPoint(p);
    update();
  }

	void PaintingWindow::drawLine(int x1, int y1, int x2, int y2,
								                const QColor& c, int penWidth)
	{
		painter_.setPen(QPen(c, penWidth));
		painter_.drawLine(x1, y1, x2, y2);
		update();
	}

  void PaintingWindow::drawLine(const QPointF& p1, const QPointF& p2, 
                                const QColor& c, int penWidth)
  {
    painter_.setPen(QPen(c, penWidth));
    painter_.drawLine(p1, p2);
    update();
  }

	void PaintingWindow::drawCircle(int xc, int yc, int r, const QColor& c,
                                  int penWidth)
  {
    painter_.setPen(QPen(c, penWidth));
    painter_.drawEllipse(QPoint(xc, yc), r, r);
    update();
  }

  void PaintingWindow::drawCircle(const QPointF& center, float r,
                                  const QColor& c, int penWidth)
  {
    painter_.setPen(QPen(c, penWidth));
    painter_.drawEllipse(QPointF(center.x(), center.y()), r, r);
    update();
  }

	void PaintingWindow::drawEllipse(int x, int y, int w, int h,
									                 const QColor& c, int penWidth)
	{
		painter_.setPen(QPen(c, penWidth));
		painter_.drawEllipse(x, y, w, h);
		update();
	}

  void PaintingWindow::drawEllipse(const QPointF& center, qreal r1, qreal r2,
                                   qreal degree, const QColor& c, int penWidth)
  {
    painter_.save();
    painter_.setPen(QPen(c, penWidth));
    painter_.translate(center);
    painter_.rotate(degree);
    painter_.translate(-r1, -r2);
    painter_.drawEllipse(QRectF(0, 0, 2*r1, 2*r2));
    painter_.restore();
    update();
  }

	void PaintingWindow::drawPoly(const QPolygonF& polygon, const QColor& c,
								                int width)
	{
		painter_.setPen(QPen(c, width));
		painter_.drawPolygon(polygon);
		update();
	}

	void PaintingWindow::drawRect(int x, int y, int w, int h, const QColor& c,
								                int penWidth)
	{
		painter_.setPen(QPen(c, penWidth));
		painter_.drawRect(x, y, w, h);
		update();
	}

	void PaintingWindow::drawText(int x,int y,const QString& s,const QColor& c,
								                int fontSize, double alpha, bool italic,
								                bool bold, bool underline)
	{
		QFont font;
		font.setPointSize(fontSize);
		font.setItalic(italic);
		font.setBold(bold);
		font.setUnderline(underline);
		painter_.setPen(c);
		painter_.setFont(font);
		painter_.rotate(qreal(alpha));
		painter_.drawText(x, y, s);
		update();
	}

	void PaintingWindow::drawArrow(int a, int b, int c, int d,
								                 const QColor& col,
								                 int arrowWidth, int arrowHeight, int style,
								                 int width)
	{
		double sl;
		double dx = c-a;
		double dy = d-b;
		double norm= qSqrt(dx*dx+dy*dy);
		if (norm < 0.999) // null vector
		{
			painter_.setPen(QPen(col, width));
			painter_.drawPoint(a, b);
			update();
			return;
		}

		QPainterPath path;
		QPolygonF pts;
		
		qreal dx_norm = dx / norm;
		qreal dy_norm = dy / norm;
		qreal p1x = a+dx_norm*(norm-arrowWidth) + arrowHeight/2.*dy_norm;
		qreal p1y = b+dy_norm*(norm-arrowWidth) - arrowHeight/2.*dx_norm;
		qreal p2x = a+dx_norm*(norm-arrowWidth) - arrowHeight/2.*dy_norm;
		qreal p2y = b+dy_norm*(norm-arrowWidth) + arrowHeight/2.*dx_norm;
		switch(style) {
			case 0:
				painter_.setPen(QPen(col, width));
				painter_.drawLine(a, b, c, d);
				painter_.drawLine(c, d, int(p1x), int(p1y));
				painter_.drawLine(c, d, int(p2x), int(p2y));
				break;
			case 1:
				pts << QPointF(p2x, p2y);
				pts << QPointF(c, d);
				pts << QPointF(p1x, p1y);
				sl = norm-(arrowWidth*.7);
				pts << QPointF(a+dx_norm*sl+dy_norm*width, 
							   b+dy_norm*sl-dx_norm*width);
				pts << QPointF(a+dy_norm*width, b-dx_norm*width);
				pts << QPointF(a-dy_norm*width, b+dx_norm*width);
				pts << QPointF(a+dx_norm*sl-dy_norm*width,
							   b+dy_norm*sl+dx_norm*width);
				path.addPolygon(pts);
				painter_.fillPath(path, col);
				break;
			case 2:
				pts << QPointF(p2x, p2y);
				pts << QPointF(c, d);
				pts << QPointF(p1x, p1y);
				sl = norm-arrowWidth;
				pts << QPointF(a+dx_norm*sl+dy_norm*width, 
							   b+dy_norm*sl-dx_norm*width);
				pts << QPointF(a+dy_norm*width, b-dx_norm*width);
				pts << QPointF(a-dy_norm*width, b+dx_norm*width);
				pts << QPointF(a+dx_norm*sl-dy_norm*width,
							   b+dy_norm*sl+dx_norm*width);
				path.addPolygon(pts);
				painter_.fillPath(path, col);
				break;
			default:
				break;
		}
		
		update();
	}

	void PaintingWindow::display(const QImage& image, int xoff, int yoff,
								               double fact)
	{
		painter_.translate(xoff, yoff);
		painter_.scale(qreal(fact), qreal(fact));
		painter_.drawImage(0, 0, image);
		painter_.scale(qreal(1./fact), qreal(1./fact));
		painter_.translate(-xoff, -yoff);
		update();
	}

	void PaintingWindow::fillCircle(int x, int y, int r, const QColor& c)
	{
		QPainterPath path;
		path.addEllipse(qreal(x)-r/2., qreal(y)-r/2., qreal(r), qreal(r));
		painter_.fillPath(path, c);
		update();
	}

  void PaintingWindow::fillCircle(const QPointF& p, qreal r, const QColor& c)
  {
    QPainterPath path;
    path.addEllipse(p, r, r);
    painter_.fillPath(path, c);
    update();
  }

	void PaintingWindow::fillEllipse(int x, int y, int w, int h, 
								                   const QColor& c)
	{
		QPainterPath path;
		path.addEllipse(qreal(x), qreal(y), qreal(w), qreal(h));
		painter_.fillPath(path, c);
		update();
	}

  void PaintingWindow::fillEllipse(const QPointF& p, qreal rx, qreal ry,
                                   qreal degree, const QColor& c)
  {
    painter_.save();
    painter_.translate(p);
    painter_.rotate(degree);
    painter_.translate(-rx, -ry);
    QPainterPath path;
    path.addEllipse(0., 0., 2*rx, 2*ry);
    painter_.fillPath(path, c);
    painter_.restore();
    update();
  }

	void PaintingWindow::fillPoly(const QPolygonF& polygon, const QColor& c)
	{
		QPainterPath path;
		path.addPolygon(polygon);
		painter_.fillPath(path, c);
		update();
	}

	void PaintingWindow::fillRect(int x, int y, int w, int h, 
								                const QColor& c)
	{
		painter_.setPen(c);
		painter_.fillRect(x, y, w, h, c);
		update();
	}

	void PaintingWindow::clear()
	{
		pixmap_.fill();
		update();
	}

  void PaintingWindow::setAntialiasing(bool on)
  { painter_.setRenderHints(QPainter::Antialiasing, on); }

  void PaintingWindow::setTransparency(bool on)
  {
    if (on)
      painter_.setCompositionMode(QPainter::CompositionMode_Multiply);
    else
      painter_.setCompositionMode(QPainter::CompositionMode_Source);
  }

  void PaintingWindow::saveScreen(const QString& filename)
  {
    pixmap_.save(filename);
  }

	void PaintingWindow::waitForEvent(int ms)
	{
		event_listening_timer_.setInterval(ms);
		event_listening_timer_.start();
	}

	void PaintingWindow::eventListeningTimerStopped()
	{
		emit sendEvent(noEvent());
	}

	void PaintingWindow::mouseMoveEvent(QMouseEvent *event)
	{
		emit(event->x(), event->y(), event->buttons());

		if (event_listening_timer_.isActive())
		{
			event_listening_timer_.stop();
			sendEvent(mouseMoved(event->x(), event->y(), event->buttons(),
				        event->modifiers()));
		}
	}

	void PaintingWindow::mousePressEvent(QMouseEvent *event)
	{
#ifdef Q_OS_MAC
		Qt::MouseButtons buttons = (event->modifiers() == Qt::ControlModifier &&
									event->buttons() == Qt::LeftButton) ? 
		Qt::MiddleButton : event->buttons();
		emit pressedMouseButtons(event->x(), event->y(), buttons);
#else
		emit pressedMouseButtons(event->x(), event->y(), event->buttons());
#endif
		if (event_listening_timer_.isActive())
		{
			event_listening_timer_.stop();
			sendEvent(mousePressed(event->x(), event->y(), event->buttons(), 
				        event->modifiers()));
		}
	}

	void PaintingWindow::mouseReleaseEvent(QMouseEvent *event)
	{
		//qDebug() << "Released " << event->pos().x() << " " << event->pos().y();
#ifdef Q_OS_MAC
		Qt::MouseButtons buttons = (event->modifiers() == Qt::ControlModifier &&
			event->button() == Qt::LeftButton) ? 
			Qt::MiddleButton : event->button();
		emit releasedMouseButtons(event->x(), event->y(), buttons);
#else
    //qDebug() << int(event->button());
		emit releasedMouseButtons(event->x(), event->y(), event->button());
#endif
		if (event_listening_timer_.isActive())
		{
			event_listening_timer_.stop();
			sendEvent(mouseReleased(event->x(), event->y(), 
					                    event->button(), event->modifiers()));
		}
	}

  void PaintingWindow::keyPressEvent(QKeyEvent *event)
  {
    emit pressedKey(event->key());
    if (event_listening_timer_.isActive())
    {
      event_listening_timer_.stop();
      sendEvent(keyPressed(event->key(), event->modifiers()));
    }
  }

  void PaintingWindow::keyReleaseEvent(QKeyEvent *event)
  {
    emit releasedKey(event->key());
    if (event_listening_timer_.isActive())
    {
      event_listening_timer_.stop();
      sendEvent(keyReleased(event->key(), event->modifiers()));
    }
  }

	void PaintingWindow::paintEvent(QPaintEvent *event)
	{		
		QPainter p(this);
		p.drawPixmap(0, 0, pixmap_);
	}

} /* namespace DO */

File: python/do/sara/graphics/derived_qobjects/painting_window.py
from PySide2.QtCore import (QObject, QPointF, QRectF, QTimer, Qt, qWarning,
                            Signal, Slot)
from PySide2.QtWidgets import QApplication, QScrollArea, QWidget
from PySide2.QtGui import QColor, QPainter, QPen, QPixmap


class ScrollArea(QScrollArea):

    def __init__(self, parent=None):
        super(ScrollArea, self).__init__(parent=parent)
        self.setAlignment(Qt.AlignCenter)
        self.setAttribute(Qt.WA_DeleteOnClose)

    def closeEvent(self, event):
        if event.spontaneous():
            qWarning((
                "\n\nWarning: you closed a window unexpectedly!\n\n"
                "Warning: the graphical application cannot terminate "
                "by design:"
                "Warning: please kill the application again manually..."
            ))


class PaintingWindowSignals(QObject):
    pressed_key = Signal(int)
    released_key = Signal(int)


class PaintingWindow(QWidget):

    def __init__(self, sizes, window_title="Sara", position=None, parent=None):
        w, h = sizes

        super(PaintingWindow, self).__init__()
        self._scroll_area = ScrollArea(parent=parent)
        self._pixmap = QPixmap(w, h)
        self._painter = QPainter(self._pixmap)
        self._painter.end()

        self._antialiasing = False

        self.setParent(self._scroll_area)
        self.setFocusPolicy(Qt.WheelFocus)

        # Populate the signals.
        self.signals = PaintingWindowSignals()

        # Set event listener.
        self._event_listening_timer = QTimer(self)
        self._event_listening_timer.setSingleShot(True)
        self._event_listening_timer.timeout.connect(
            PaintingWindow.eventListeningTimerStopped
        )

        # Move widget.
        if position is not None:
            self._scroll_area.move(position[0], position[1])
        self._scroll_area.setWindowTitle(window_title)
        self._scroll_area.setWidget(self)
        self._scroll_area.setFocusProxy(self)

        # Maximize if necessary.
        if w >= QApplication.instance().desktop().width() or \
                h >= QApplication.instance().desktop().height():
            self._scroll_area.showMaximized()
        # Resize the scroll area with the size plus a two-pixel offset.
        else:
            self._scroll_area.resize(w + 2, h + 2)
        self.resize(w, h)

        # Initialize the pixmap.
        self._pixmap.fill()
        self.update()

        # Show the widget.
        self._scroll_area.show()

    @property
    def window_title(self):
        return self._scroll_area.windowTitle()

    @property
    def x(self):
        return self._scroll_area.pos().x();

    @property
    def y(self):
        return self._scroll_area.pos().y();

    def keyPressEvent(self, event):
        self.signals.pressed_key.emit(event.key())
        # if (m_eventListeningTimer.isActive())
        # {
        #     m_eventListeningTimer.stop();
        #     emit sendEvent(key_pressed(event->key(), event->modifiers()));
        # }

    def keyReleaseEvent(self, event):
        self.signals.released_key.emit(event.key());
        # {
        #   if (m_eventListeningTimer.isActive())
        #   {
        #     m_eventListeningTimer.stop();
        #     emit sendEvent(key_released(event->key(), event->modifiers()));
        #   }
        # }

    @Slot()
    def eventListeningTimerStopped(self):
        pass

    def paintEvent(self, event):
        p = QPainter(self)
        p.drawPixmap(0, 0, self._pixmap)

    def draw_point(self, x, y, color):
        self._painter.begin(self._pixmap)
        self._painter.setRenderHints(QPainter.Antialiasing, self._antialiasing);
        self._painter.setPen(QColor(*color));
        self._painter.drawPoint(x, y);
        self._painter.end()
        self.update()

    def draw_line(self, x1, y1, x2, y2, color, pen_width):
        self._painter.begin(self._pixmap)
        self._painter.setRenderHints(QPainter.Antialiasing, self._antialiasing);
        self._painter.setPen(QPen(QColor(*color), pen_width))
        self._painter.drawLine(x1, y1, x2, y2)
        self._painter.end()
        self.update()

    def draw_image(self, image, offset, scale):
        xoff, yoff = offset
        self._painter.begin(self._pixmap)
        self._painter.setRenderHints(QPainter.Antialiasing, self._antialiasing);
        self._painter.translate(xoff, yoff)
        self._painter.scale(scale, scale)
        self._painter.drawImage(0, 0, image)
        self._painter.scale(1 / scale, 1 / scale)
        self._painter.translate(-xoff, -yoff)
        self._painter.end()
        self.update()

    def draw_circle(self, center, radius, color, pen_width):
        self._painter.begin(self._pixmap)
        self._painter.setRenderHints(QPainter.Antialiasing, self._antialiasing);
        self._painter.setPen(QColor(*color))
        self._painter.drawEllipse(QPointF(*center), radius, radius);
        self._painter.end()
        self.update()

    def draw_ellipse(self, center, r1, r2, angle_in_degrees, color, pen_width):
        self._painter.begin(self._pixmap)
        self._painter.setRenderHints(QPainter.Antialiasing, self._antialiasing);
        self._painter.save()
        self._painter.setPen(QColor(*color))
        self._painter.translate(QPointF(*center))
        self._painter.rotate(angle_in_degrees)
        self._painter.translate(-r1, -r2)
        self._painter.drawEllipse(QRectF(0, 0, 2 * r1, 2 * r2));
        self._painter.restore()
        self._painter.end()
        self.update()

    def clear(self):
        self._pixmap.fill();
        self.update();

    def set_antialiasing(self, on):
        self._antialiasing = on

#  void PaintingWindow::drawPoly(const QPolygonF& polygon, const QColor& c,
#                                int width)
#  {
#    m_painter.setPen(QPen(c, width));
#    m_painter.drawPolygon(polygon);
#    update();
#  }
#
#  void PaintingWindow::drawRect(int x, int y, int w, int h, const QColor& c,
#                                int penWidth)
#  {
#    m_painter.setPen(QPen(c, penWidth));
#    m_painter.drawRect(x, y, w, h);
#    update();
#  }
#
#  void PaintingWindow::drawText(int x, int y, const QString& text,
#                                const QColor& color, int fontSize,
#                                double orientation, bool italic, bool bold,
#                                bool underline)
#  {
#    QFont font;
#    font.setPointSize(fontSize);
#    font.setItalic(italic);
#    font.setBold(bold);
#    font.setUnderline(underline);
#
#    m_painter.save();
#    m_painter.setPen(color);
#    m_painter.setFont(font);
#
#    m_painter.translate(x, y);
#    m_painter.rotate(qreal(orientation));
#    m_painter.drawText(0, 0, text);
#    m_painter.restore();
#    update();
#  }
#
#  void PaintingWindow::drawArrow(int x1, int y1, int x2, int y2,
#                                 const QColor& col,
#                                 int arrowWidth, int arrowHeight, int style,
#                                 int width)
#  {
#    double sl;
#    double dx = x2-x1;
#    double dy = y2-y1;
#    double norm= qSqrt(dx*dx+dy*dy);
#    if (norm < 0.999) // null vector
#    {
#      m_painter.setPen(QPen(col, width));
#      m_painter.drawPoint(x1, y1);
#      update();
#      return;
#    }
#
#    QPainterPath path;
#    QPolygonF pts;
#
#    qreal dx_norm = dx / norm;
#    qreal dy_norm = dy / norm;
#    qreal p1x = x1 + dx_norm*(norm-arrowWidth) + arrowHeight/2.*dy_norm;
#    qreal p1y = y1 + dy_norm*(norm-arrowWidth) - arrowHeight/2.*dx_norm;
#    qreal p2x = x1 + dx_norm*(norm-arrowWidth) - arrowHeight/2.*dy_norm;
#    qreal p2y = y1 + dy_norm*(norm-arrowWidth) + arrowHeight/2.*dx_norm;
#    switch(style) {
#      case 0:
#        m_painter.setPen(QPen(col, width));
#        m_painter.drawLine(x1, y1, x2, y2);
#        m_painter.drawLine(x2, y2, int(p1x), int(p1y));
#        m_painter.drawLine(x2, y2, int(p2x), int(p2y));
#        break;
#      case 1:
#        pts << QPointF(p2x, p2y);
#        pts << QPointF(x2, y2);
#        pts << QPointF(p1x, p1y);
#        sl = norm-(arrowWidth*.7);
#        pts << QPointF(x1 + dx_norm*sl + dy_norm*width,
#                       y1 + dy_norm*sl - dx_norm*width);
#        pts << QPointF(x1 + dy_norm*width, y1 - dx_norm*width);
#        pts << QPointF(x1 - dy_norm*width, y1 + dx_norm*width);
#        pts << QPointF(x1 + dx_norm*sl - dy_norm*width,
#                       y1 + dy_norm*sl + dx_norm*width);
#        path.addPolygon(pts);
#        m_painter.fillPath(path, col);
#        break;
#      case 2:
#        pts << QPointF(p2x, p2y);
#        pts << QPointF(x2, y2);
#        pts << QPointF(p1x, p1y);
#        sl = norm-arrowWidth;
#        pts << QPointF(x1 + dx_norm*sl + dy_norm*width,
#                       y1 + dy_norm*sl - dx_norm*width);
#        pts << QPointF(x1 + dy_norm*width, y1-dx_norm*width);
#        pts << QPointF(x1 - dy_norm*width, y1+dx_norm*width);
#        pts << QPointF(x1 + dx_norm*sl - dy_norm*width,
#                       y1 + dy_norm*sl + dx_norm*width);
#        path.addPolygon(pts);
#        m_painter.fillPath(path, col);
#        break;
#      default:
#        break;
#    }
#
#    update();
#  }
#
#  void PaintingWindow::fillCircle(int x, int y, int r, const QColor& c)
#  {
#    QPainterPath path;
#    path.addEllipse(qreal(x)-r/2., qreal(y)-r/2., qreal(r), qreal(r));
#    m_painter.fillPath(path, c);
#    update();
#  }
#
#  void PaintingWindow::fillCircle(const QPointF& p, qreal r, const QColor& c)
#  {
#    QPainterPath path;
#    path.addEllipse(p, r, r);
#    m_painter.fillPath(path, c);
#    update();
#  }
#
#  void PaintingWindow::fillEllipse(int x, int y, int w, int h,
#                                   const QColor& c)
#  {
#    QPainterPath path;
#    path.addEllipse(qreal(x), qreal(y), qreal(w), qreal(h));
#    m_painter.fillPath(path, c);
#    update();
#  }
#
#  void PaintingWindow::fillEllipse(const QPointF& p, qreal rx, qreal ry,
#                                   qreal degree, const QColor& c)
#  {
#    m_painter.save();
#    m_painter.translate(p);
#    m_painter.rotate(degree);
#    m_painter.translate(-rx, -ry);
#    QPainterPath path;
#    path.addEllipse(0., 0., 2*rx, 2*ry);
#    m_painter.fillPath(path, c);
#    m_painter.restore();
#    update();
#  }
#
#  void PaintingWindow::fillPoly(const QPolygonF& polygon, const QColor& c)
#  {
#    QPainterPath path;
#    path.addPolygon(polygon);
#    m_painter.fillPath(path, c);
#    update();
#  }
#
#  void PaintingWindow::fillRect(int x, int y, int w, int h,
#                                const QColor& c)
#  {
#    m_painter.setPen(c);
#    m_painter.fillRect(x, y, w, h, c);
#    update();
#  }
#
#  void PaintingWindow::setTransparency(bool on)
#  {
#    if (on)
#      m_painter.setCompositionMode(QPainter::CompositionMode_Multiply);
#    else
#      m_painter.setCompositionMode(QPainter::CompositionMode_Source);
#  }
#
#  void PaintingWindow::saveScreen(const QString& filename)
#  {
#    m_pixmap.save(filename);
#  }
#
#  void PaintingWindow::resizeScreen(int width, int height)
#  {
#    if (m_pixmap.width() == width && m_pixmap.height() == height)
#      return;
#    /*
#       The following internal changes are critical to prevent Qt from crashing.
#       1. Tell QPainter 'painter_' to stop using using QPixmap 'pixmap_'.
#       2. Reinitialize the QPixmap with the new size.
#       3. Now we can re-allow QPainter 'painter_' to re-use QPixmap 'pixmap_'.
#     */
#    m_painter.end();
#    m_pixmap = QPixmap(width, height);
#    m_pixmap.fill();
#    m_painter.begin(&m_pixmap);
#
#    // Resize the window and the scroll area as follows.
#    resize(width, height);
#    if (width > qApp->desktop()->width() || height > qApp->desktop()->height())
#    {
#      width = 800;
#      height = 600;
#    }
#    self._scroll_area->resize(width+2, height+2);
#  }
#
#  void PaintingWindow::waitForEvent(int ms)
#  {
#    m_eventListeningTimer.setInterval(ms);
#    m_eventListeningTimer.start();
#  }
#
#  void PaintingWindow::eventListeningTimerStopped()
#  {
#    emit sendEvent(no_event());
#  }
#
#  void PaintingWindow::mouseMoveEvent(QMouseEvent *event)
#  {
#    emit movedMouse(event->x(), event->y(), event->buttons());
#
#    if (m_eventListeningTimer.isActive())
#    {
#      m_eventListeningTimer.stop();
#      emit sendEvent(mouse_moved(event->x(), event->y(), event->buttons(),
#                     event->modifiers()));
#    }
#  }
#
#  void PaintingWindow::mousePressEvent(QMouseEvent *event)
#  {
##ifdef Q_OS_MAC
#    Qt::MouseButtons buttons = (event->modifiers() == Qt::ControlModifier &&
#                  event->buttons() == Qt::LeftButton) ?
#    Qt::MiddleButton : event->buttons();
#    emit pressedMouseButtons(event->x(), event->y(), buttons);
##else
#    emit pressedMouseButtons(event->x(), event->y(), event->buttons());
##endif
#    if (m_eventListeningTimer.isActive())
#    {
#      m_eventListeningTimer.stop();
#      emit sendEvent(mouse_pressed(event->x(), event->y(), event->buttons(),
#                     event->modifiers()));
#    }
#  }
#
#  void PaintingWindow::mouseReleaseEvent(QMouseEvent *event)
#  {
##ifdef Q_OS_MAC
#    Qt::MouseButtons buttons = (event->modifiers() == Qt::ControlModifier &&
#                                event->buttons() == Qt::LeftButton) ?
#      Qt::MiddleButton : event->buttons();
#    emit releasedMouseButtons(event->x(), event->y(), buttons);
##else
#    emit releasedMouseButtons(event->x(), event->y(), event->button());
##endif
#    if (m_eventListeningTimer.isActive())
#    {
#      m_eventListeningTimer.stop();
#      emit sendEvent(mouse_released(event->x(), event->y(),
#                                   event->buttons(), event->modifiers()));
#    }
#  }
